File: src/connectors/dns_native.js
import { BaseConnector } from './base.js';
import dns from 'dns/promises';

export class DnsNativeConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'DNS-Native';
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        const entities = [];
        try {
            const recordTypes = ['A', 'AAAA', 'MX', 'NS', 'TXT', 'CNAME', 'SOA', 'SRV'];

            for (const type of recordTypes) {
                try {
                    const records = await dns.resolve(inputValue, type);
                    for (const rec of (Array.isArray(records) ? records : [records])) {
                        if (type === 'A' || type === 'AAAA') {
                            entities.push({
                                type: 'ip',
                                data: { address: rec, version: type === 'AAAA' ? 6 : 4, source: 'dns_resolve' },
                                confidence: 0.95,
                                asKnown: { type: 'ip', value: rec }
                            });
                            entities.push({
                                type: 'dns_record',
                                data: { type, name: inputValue, value: rec },
                                confidence: 0.95
                            });
                        } else if (type === 'MX') {
                            entities.push({
                                type: 'dns_record',
                                data: { type: 'MX', name: inputValue, value: rec.exchange, priority: rec.priority },
                                confidence: 0.95
                            });
                            entities.push({
                                type: 'subdomain',
                                data: { name: rec.exchange, parent_domain: inputValue, source: 'mx_record' },
                                confidence: 0.85
                            });
                        } else if (type === 'NS') {
                            entities.push({
                                type: 'dns_record',
                                data: { type: 'NS', name: inputValue, value: rec },
                                confidence: 0.95
                            });
                        } else if (type === 'TXT') {
                            const txt = Array.isArray(rec) ? rec.join('') : rec;
                            entities.push({
                                type: 'dns_record',
                                data: { type: 'TXT', name: inputValue, value: txt },
                                confidence: 0.95
                            });

                            const spfIncludes = txt.match(/include:(\S+)/g);
                            if (spfIncludes) {
                                for (const inc of spfIncludes) {
                                    const domain = inc.replace('include:', '');
                                    entities.push({
                                        type: 'domain',
                                        data: { name: domain, source: 'spf_include' },
                                        confidence: 0.7,
                                        asKnown: { type: 'domain', value: domain }
                                    });
                                }
                            }

                            const dmarcMatch = txt.match(/rua=mailto:([^;,\s]+)/);
                            if (dmarcMatch) {
                                entities.push({
                                    type: 'email',
                                    data: { address: dmarcMatch[1], source: 'dmarc_rua' },
                                    confidence: 0.8,
                                    asKnown: { type: 'email', value: dmarcMatch[1] }
                                });
                            }

                            if (txt.includes('v=spf1')) {
                                const ip4s = txt.match(/ip4:(\S+)/g);
                                if (ip4s) {
                                    for (const ip4 of ip4s) {
                                        const ip = ip4.replace('ip4:', '').split('/')[0];
                                        entities.push({
                                            type: 'ip',
                                            data: { address: ip, source: 'spf_record' },
                                            confidence: 0.8,
                                            asKnown: { type: 'ip', value: ip }
                                        });
                                    }
                                }
                            }
                        } else if (type === 'CNAME') {
                            entities.push({
                                type: 'dns_record',
                                data: { type: 'CNAME', name: inputValue, value: rec },
                                confidence: 0.95
                            });
                            entities.push({
                                type: 'domain',
                                data: { name: rec, source: 'cname' },
                                confidence: 0.8,
                                asKnown: { type: 'domain', value: rec }
                            });
                        } else if (type === 'SOA') {
                            entities.push({
                                type: 'dns_record',
                                data: {
                                    type: 'SOA', name: inputValue,
                                    value: `${rec.nsname} ${rec.hostmaster}`,
                                    serial: rec.serial
                                },
                                confidence: 0.95
                            });
                            if (rec.hostmaster) {
                                const parts = rec.hostmaster.split('.');
                                const email = parts.length >= 2
                                    ? parts[0] + '@' + parts.slice(1).join('.')
                                    : rec.hostmaster;
                                entities.push({
                                    type: 'email',
                                    data: { address: email, source: 'soa_hostmaster' },
                                    confidence: 0.5
                                });
                            }
                        }
                    }
                } catch {}
            }

            const commonSubs = ['www', 'mail', 'ftp', 'api', 'dev', 'staging', 'admin', 'blog', 'shop', 'store',
                'app', 'cdn', 'static', 'media', 'img', 'images', 'vpn', 'remote', 'portal', 'login',
                'dashboard', 'status', 'docs', 'help', 'support', 'git', 'gitlab', 'jira', 'wiki',
                'test', 'beta', 'demo', 'sandbox', 'internal', 'intranet', 'mx', 'smtp', 'pop', 'imap',
                'ns1', 'ns2', 'dns', 'proxy', 'gateway', 'firewall', 'monitor', 'grafana', 'prometheus'];

            const subResults = await Promise.allSettled(
                commonSubs.map(async sub => {
                    const fqdn = `${sub}.${inputValue}`;
                    const addrs = await dns.resolve4(fqdn);
                    return { sub, fqdn, ips: addrs };
                })
            );

            for (const result of subResults) {
                if (result.status === 'fulfilled' && result.value.ips?.length) {
                    const { fqdn, ips } = result.value;
                    entities.push({
                        type: 'subdomain',
                        data: { name: fqdn, parent_domain: inputValue, ip: ips[0], source: 'dns_bruteforce' },
                        confidence: 0.9,
                        asKnown: { type: 'domain', value: fqdn }
                    });
                    for (const ip of ips) {
                        entities.push({
                            type: 'ip',
                            data: { address: ip, source: 'subdomain_resolve' },
                            confidence: 0.9,
                            asKnown: { type: 'ip', value: ip }
                        });
                    }
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence || 0.5);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    async healthCheck() { return { ok: true, version: 'native' }; }
}

File: cyclops/connectors/dns_native.py
import asyncio
import socket
from cyclops.connectors.base import BaseConnector


class DnsNativeConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "DNS-Native"

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})

        try:
            entities = []
            loop = asyncio.get_event_loop()

            record_types = ["A", "AAAA", "MX", "NS", "TXT", "SOA", "CNAME"]
            for rtype in record_types:
                try:
                    answers = await loop.run_in_executor(None, lambda rt=rtype: self._resolve(input_value, rt))
                    for answer in answers:
                        entities.append({
                            "type": "dns_record",
                            "data": {"type": rtype, "name": input_value, "value": answer, "source": "dns_native"},
                            "confidence": 0.95,
                        })
                        if rtype in ("A", "AAAA"):
                            entities.append({
                                "type": "ip",
                                "data": {"address": answer, "source": "dns_native"},
                                "confidence": 0.95,
                                "as_known": {"type": "ip", "value": answer},
                            })
                        elif rtype == "MX":
                            mx_host = answer.split()[-1].rstrip(".")
                            if mx_host:
                                entities.append({
                                    "type": "subdomain",
                                    "data": {"name": mx_host, "parent_domain": input_value, "source": "dns_mx"},
                                    "confidence": 0.9,
                                })
                        elif rtype == "NS":
                            ns_host = answer.rstrip(".")
                            entities.append({
                                "type": "subdomain",
                                "data": {"name": ns_host, "source": "dns_ns"},
                                "confidence": 0.9,
                            })
                        elif rtype == "TXT":
                            if "v=spf1" in answer:
                                for part in answer.split():
                                    if part.startswith("include:"):
                                        entities.append({
                                            "type": "domain",
                                            "data": {"name": part[8:], "source": "dns_spf"},
                                            "confidence": 0.7,
                                        })
                        elif rtype == "SOA":
                            parts = answer.split()
                            if len(parts) >= 2:
                                hostmaster = parts[1].rstrip(".")
                                hp = hostmaster.split(".")
                                if len(hp) >= 2:
                                    email = hp[0] + "@" + ".".join(hp[1:])
                                    entities.append({
                                        "type": "email",
                                        "data": {"address": email, "source": "soa_hostmaster"},
                                        "confidence": 0.5,
                                    })
                except Exception:
                    continue

            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)

            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    def _resolve(self, domain, record_type):
        import subprocess
        try:
            result = subprocess.run(
                ["dig", "+short", domain, record_type],
                capture_output=True, text=True, timeout=10,
            )
            return [line.strip() for line in result.stdout.strip().split("\n") if line.strip()]
        except Exception:
            if record_type == "A":
                try:
                    return [r[4][0] for r in socket.getaddrinfo(domain, None, socket.AF_INET)]
                except Exception:
                    return []
            return []

    async def health_check(self):
        return {"ok": True, "version": "native"}
